### PR TITLE
eds-caching: introduce the EDS caching into EdsClusterImpl

### LINKED ISSUE #26749: Using cached EDS resources
*Title*: *Using cached EDS resources*

*Description*:
Currently after an EDS-cluster update, Envoy waits for an EDS response. If a timeout occurs, the EDS-cluster will be used with an empty assignment. This may break ongoing traffic, if the xDS server fails to reply in a timely manner. As noted in #13009 this can be solved if Envoy will cache the EDS resources.

Unfortunately Envoy cannot immediately use the cached resource because there are scenarios where the Cluster update requires an updated ClusterLoadAssignment (CLA), and using the cached resource will break traffic. An example is when a non-TLS cluster is upgraded to supporting TLS (see #11877, and [this test](https://github.com/envoyproxy/envoy/blob/25c0e7c058d74772cc2340cb3e726adced0dc889/test/integration/ads_integration_test.cc#L560-L613)) and using the previously assigned CLA will break traffic. A possible way to solve this while still adhering to the xDS protocol is to require that upon a material change to the Cluster the xDS-server will also modify the [Eds `service_name`]https://github.com/envoyproxy/envoy/blob/25c0e7c058d74772cc2340cb3e726adced0dc889/api/envoy/config/cluster/v3/cluster.proto#L207). However, many servers do not adhere to this, and to ensure backwards compatibility, a different approach is needed.

*Suggested approach*
The idea is to allow fallback to a cached EDS resource as follows:
1. When receiving a new/updated EDS-Cluster, wait for a ClusterLoadAssignment (CLA) for that cluster (Cluster is in a warming state).
2. If the CLA doesn't arrive within a certain timeframe, fetch the CLA resource from the cache, if exists, and make the Cluster active.
3. Otherwise, make the Cluster active with an empty assignment.

Note that this doesn't introduce a new way to break traffic, just prevents breaking traffic in the case where the CLA isn't modified and/or isn't sent from the xDS server. If a material change to the cluster occurs, but an updated assignment isn't received, traffic will be broken, just as happens today.

*Plan*
1. Introduce a cache for EDS resources.
2. Plumb the cache into the [ClusterManagerFactory](https://github.com/envoyproxy/envoy/blob/25c0e7c058d74772cc2340cb3e726adced0dc889/envoy/upstream/cluster_manager.h#L467).
3. Update [EdsClusterImpl](https://github.com/envoyproxy/envoy/blob/25c0e7c058d74772cc2340cb3e726adced0dc889/source/extensions/clusters/eds/eds.h#L29) to:
    * Store a resource in the cache.
    * Fetch the resource upon timeout.
    * Ensure that the resource TTL is enforced for cached resources.

*Special considerations*
The solution will need to clarify the lifetime of the cached resources.
The solution will target EDS resources only, as other resource (such as RDS) are currently "cached" and adhere to the protocol. In the future, the cache may be used for other types of resources, if needed.

*Other implications*
As with any cache, this will require additional memory. An additional O(N) memory will be required, where N is the number of CLA resources (or EDS clusters). Note that the resource is only accessed by the main thread (no copies to worker threads are needed).

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: eds-caching: introduce the EDS caching into EdsClusterImpl
Additional Description:
This is the last PR to close #26749.
In this PR the EdsClusterImpl is modified to use the cached resource when a warming timeout occurs.
The PR only includes support for EDS caching when ADS is used.
The runtime guard `envoy.restart_features.use_eds_cache_for_ads` was introduced for backward compatibility.
Although this change modifies Envoy's behavior with EDS, the intention is to not breaking the current behavior (e.g., modifying a cluster from non-TLS to TLS will still work as it did previously, see #26749 for more information).
The cache will incur more memory.

EDIT: Following an internal discussion, the runtime guard is set to false by default to allow gradual rollout of this feature, and testing it in production.

Risk Level: Medium - changes behavior of EDS over ADS.
Testing: Added unit and integration tests.
Docs Changes: N/A.
Release Notes: Added.
Platform Specific Features: N/A.

### CODE REVIEW

File: changelogs/current.yaml
Comment (adisuissa) [inline changelogs/current.yaml:9]: Hmm... CI complains that the doc label is undefined: https://dev.azure.com/cncf/envoy/_build/results?buildId=145343&view=logs&j=9fd58ae0-0c50-56b1-f517-c2b9afda9c9b&t=c8fc762c-55bb-57d4-4cab-8ec527e7a423&l=117.

@phlax not sure why this is so, IIUC the [field](https://github.com/envoyproxy/envoy/blob/35684d731a2efc5a3415be191e8a8944539ccf4f/api/envoy/config/core/v3/config_source.proto#L248C3-L248C3) is correctly referenced. Am I missing anything?

Comment (phlax) [reply changelogs/current.yaml:9]: i can see 

```rst
.. _envoy_v3_api_field_config.core.v3.ConfigSource.initial_fetch_timeout:
```

here https://www.envoyproxy.io/docs/envoy/latest/_sources/api-v3/config/core/v3/config_source.proto.rst.txt

so ref looks correct

Comment (phlax) [inline changelogs/current.yaml:9]:
```suggestion
    :ref:`initial_fetch_timeout <envoy_v3_api_field_config.core.v3.ConfigSource.initial_fetch_timeout>` times out, and then apply
```

typo

Comment (adisuissa) [reply changelogs/current.yaml]: Oops :face_palm:, thanks for pointing it out!

File: source/extensions/clusters/eds/eds.cc
Comment (abeyad) [inline source/extensions/clusters/eds/eds.cc:328]: s/time/timer

File: source/extensions/clusters/eds/eds.h
Comment (abeyad) [inline source/extensions/clusters/eds/eds.h:67]: nit: dash not required in data-structures

File: test/extensions/clusters/eds/eds_test.cc
Comment (abeyad) [inline test/extensions/clusters/eds/eds_test.cc:3173]: Love the tests in this file and they're very clearly documented!

File: source/extensions/clusters/eds/eds.cc
Comment (abeyad) [inline source/extensions/clusters/eds/eds.cc:425]: Question: what is the `initial_fetch_timeout` typically specified to?  The default seems to be 15s.  This means that if the management server wants to save on bandwidth/data exchange, then it won't send the EDS assignment updates if they haven't changed, but if the `intial_fetch_timeout` is long, then the Envoy client will just be hanging until the timeout before it can warm the cluster.  That could result in an abnormally long delay.  Am I understanding this trade-off correctly?

If so, what about the alternative of always warming with cached assignments, if available, and if the management server sends assignment updates after that, the cluster can just be updated with those assignments and proceed.

**Review by abeyad — COMMENTED** [review]
Thanks Adi, this looks great!  Left a few minor comments, and one question.

File: changelogs/current.yaml
Comment (adisuissa) [reply changelogs/current.yaml]: I agree that the comment wasn't cleared.
I can answer your question, but I want to make the comment explain it :)
LMK if I need to improve it further

File: changelogs/current.yaml
Comment (adisuissa) [reply changelogs/current.yaml]: In general, most runtime flags are true by default, and will be removed in ~2 releases.
I tried rephrasing the comment, and I hope it makes it clearer.

File: envoy/upstream/cluster_manager.h
Comment (adisuissa) [reply envoy/upstream/cluster_manager.h]: Done.

File: source/common/upstream/cluster_manager_impl.cc
Comment (adisuissa) [reply source/common/upstream/cluster_manager_impl.cc]: Done.

File: source/extensions/clusters/eds/eds.cc
Comment (adisuissa) [reply source/extensions/clusters/eds/eds.cc]: Done.

File: source/extensions/clusters/eds/eds.h
Comment (adisuissa) [reply source/extensions/clusters/eds/eds.h]: Done.

File: source/extensions/clusters/eds/eds.cc
Comment (adisuissa) [reply source/extensions/clusters/eds/eds.cc:425]: That's a good question, and I tried to give more context in the original issue (#26749).

The `initial_fetch_timeout` tells Envoy how long to wait before changing a resource from "warming" state to "active". One thing to note about Envoy's implementation is that this value is used not only for the "initial", but also for updates.

The problem is for coupled resources that are meant to be changed "together" but the protocol doesn't really allow for it. For example, an EDS cluster and its assignment may need to be updated together.
This [test](https://github.com/envoyproxy/envoy/blob/25c0e7c058d74772cc2340cb3e726adced0dc889/test/integration/ads_integration_test.cc#L560-L613) for example, was introduced to ensure that a non-TLS cluster that has been modified to use TLS, will only be active after its assignment was fetched. this is how Envoy works at the moment.

Note that this is a place where Envoy's implementation doesn't follow the xDS-protocol. The protocol says that the client should use the cached object (if it has it) immediately. However, this will break current use-cases of Envoy, like the ones that do non-TLS -> TLS upgrades.

The current solution is somewhat in-between: use the cached object after a timeout. In case an "upgrade" occurs, Envoy will wait for the assignment before it makes the cluster active. If no update comes, Envoy will "eventually" (timeout) use the cached object.

LMK if you'd like more context here, or if you think of a better approach.

Comment (adisuissa) [reply source/extensions/clusters/eds/eds.cc:425]: > Thanks for that explanation Adi. That clarifies it a lot.
> 
> My concern with using the `initial_fetch_timeout` is that you could have a cluster remain in the warmed state for quite a while (~15s if default values are used). I suppose this would occur if the xDS management server decides to _not_ send EDS assignment updates if they haven't changed.
> 
Or if the xDS management server went down before sending the EDS response. In case the assignment hasn't changed, the updated cluster will now use the cached assignment (rather than an empty assignment previously).

> IIUC, the non-TLS => TLS cluster update use case breaks if we immediately use cached EDS assignments b/c the cached assignments may be pointing to endpoints that don't support the required certificates for TLS? Or is it because the active connections will be maintained without new assignments, so we'll be sending data over non-TLs connections? Or is it something else?
> 
Yes. Essentially it is expected that the cluster will need a different set of endpoints.

> Just a couple (probably not-well-thought-out) ideas for your consideration:
> 
> 1. Should a different, new value be used for the timeout for which the default timeout is shorter (say, 2s?)?
> 2. Or should we change the `initial_fetch_timeout` default from 15s to something shorter?  Looks like we can do that in code here: https://github.com/envoyproxy/envoy/blob/7fb3f62b9edceacbce37c90b1e1e26b06abbe370/source/common/config/utility.cc#L198
>     (note: I sent [config: Update the comment on initial_fetch_timeout default value #28930](https://github.com/envoyproxy/envoy/pull/28930) to correct the comment on the default value in the header file).

I'll reach you offline to discuss this, and maybe come up with other ideas.

**Review by abeyad — APPROVED** [review]

**Comment by adisuissa** [comment]
Following an internal discussion, the runtime guard is set to false by default to allow gradual rollout of this feature, and testing it in production.

Assigning Harvey for reviewing/suggesting a senior maintainer.
/assign @htuch

**Review by abeyad — DISMISSED** [review]

File: changelogs/current.yaml
Comment (htuch) [inline changelogs/current.yaml:9]: Nit: s/upadate/update/

File: changelogs/current.yaml
Comment (htuch) [inline changelogs/current.yaml:5]: Should https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#resource-warming also be updated?

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo small doc nit / comment.

File: changelogs/current.yaml
Comment (adisuissa) [reply changelogs/current.yaml]: Done. Also updated the comment to reflect that the flag is disabled by default.

File: changelogs/current.yaml
Comment (adisuissa) [reply changelogs/current.yaml:5]: Good idea, thanks. Done.

**Pull request merged**
